ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos.hs

```diff
@@ -0,0 +1,364 @@
+  } deriving Generic
+
+instance NoUnexpectedThunks (TPraosLedgerView c)
+
+{-------------------------------------------------------------------------------
+  Protocol proper
+-------------------------------------------------------------------------------}
+
+data TPraos c
+
+-- | TPraos parameters that are node independent
+data TPraosParams = TPraosParams {
+      -- | Active slots coefficient. This parameter represents the proportion of
+      -- slots in which blocks should be issued. This can be interpreted as the
+      -- probability that a party holding all the stake will be elected as
+      -- leader for a given slot.
+      tpraosLeaderF       :: Double
+    , tpraosSecurityParam :: SecurityParam
+    } deriving Generic
+
+instance NoUnexpectedThunks TPraosParams
+
+data TPraosIsCoreNode c = TPraosIsCoreNode
+  { -- | Online KES key used to sign blocks.GHC.Generics
+    tpraosIsCoreNodeSKSHot :: SignKeyKES (TPraosKES c)
+    -- | Certificate delegating rights from the stake pool cold key (or genesis
+    -- stakeholder delegate cold key) to the online KES key.
+  , tpraosIsCoreNodeOpCert :: OCert (TPraosDSIGN c) (TPraosKES c)
+  } deriving Generic
+
+instance
+  ( DSIGNAlgorithm (TPraosDSIGN c)
+  , KESAlgorithm (TPraosKES c)
+  ) => NoUnexpectedThunks (TPraosIsCoreNode c)
+
+instance TPraosCrypto c => OuroborosTag (TPraos c) where
+  data NodeConfig (TPraos c) = TPraosNodeConfig
+    { tpraosParams        :: TPraosParams
+    , tpraosSignKeyVRF    :: SignKeyVRF (TPraosVRF c)
+    } deriving Generic
+
+
+  protocolSecurityParam = tpraosSecurityParam . tpraosParams
+
+  type NodeState       (TPraos c) = Maybe (TPraosIsCoreNode c)
+  type LedgerView      (TPraos c) = TPraosLedgerView c
+  type IsLeader        (TPraos c) = TPraosProof c
+  type ValidationErr   (TPraos c) = [[STS.PredicateFailure (PRTCL c)]]
+  type SupportedHeader (TPraos c) = HeaderSupportsTPraos c
+  type ChainState      (TPraos c) = ChainState.TPraosChainState c
+
+  checkIsLeader cfg@TPraosNodeConfig{..} slot lv cs =
+    getNodeState >>= \case
+        Nothing -> return Nothing
+        Just icn@TPraosIsCoreNode{ tpraosIsCoreNodeOpCert} -> do
+          let mkSeed' = mkSeed @(TPraosHash c) @(TPraosDSIGN c) @(TPraosKES c) @(TPraosVRF c)
+              vkhCold = hashKey $ ocertVkCold tpraosIsCoreNodeOpCert
+              t = leaderThreshold cfg lv vkhCold
+              eta0 = tpraosLedgerViewEpochNonce lv
+              prevHash = prtclStateHash @c $ ChainState.toPRTCLState cs
+              rho' = mkSeed' seedEta (convertSlot slot) eta0 prevHash
+              y' = mkSeed' seedL (convertSlot slot) eta0 prevHash
+          rho <- evalCertified rho' tpraosSignKeyVRF
+          y   <- evalCertified y'   tpraosSignKeyVRF
+          -- First, check whether we're in the overlay schedule
+          case (Map.lookup (convertSlot slot) $ tpraosLedgerViewOverlaySchedule lv) of
```

Comment:
Change the inner type of the overlay schedule to be a strict triple kind of thing